icicle-compiler/test/Icicle/Test/Sea/Zebra.hs

```diff
@@ -0,0 +1,449 @@
+                          VRight . snd <$> peekOutput buf 0 ty
+                        e ->
+                          pure (VLeft (VError e))
+
+                     peekInputs (x:xs) (index - 1)
+
+                 inputs <- peekInputs [] nfacts
+                 liftIO $ free buf
+                 return $ facts === List.reverse inputs
+              )
+        )
+
+--------------------------------------------------------------------------------
+
+data TestError
+  = ZebraError (TableError Schema)
+  | UnexpectedError ValType [BaseValue]
+  deriving (Show)
+
+data ZebraWellTyped = ZebraWellTyped {
+    zWelltyped    :: WellTyped
+  , zFactType     :: ValType -- wtFactType = Sum Error FactType
+  , zEntity       :: Zebra.Entity Schema
+  }
+
+instance Show ZebraWellTyped where
+  show (ZebraWellTyped wt _ e) =
+    pp wt e
+
+zebra :: Jack (Maybe ZebraWellTyped)
+zebra = do
+  let notArray x = case wtFactType x of
+        SumT _ (ArrayT _) -> False
+        _ -> True
+  wt <- arbitrary `suchThat` notArray
+  zebraOfWellTyped wt
+
+zebraOfWellTyped :: WellTyped -> Jack (Maybe ZebraWellTyped)
+zebraOfWellTyped wt =
+  case zebraOfFacts (wtFactType wt) (fmap atFact (wtFacts wt)) of
+    Left e ->
+      Savage.error (show e)
+    Right Nothing ->
+      return Nothing
+    Right (Just (ty, tombstones, table)) -> do
+      -- FIXME ignoring fact times for now, but to test it we should convert icicle time to 1600 epoch secs here
+      -- let ts = fmap (Zebra.Time . fromIntegral . Icicle.secondsCountJulian . atTime) (wtFacts wt)
+      let ts = List.replicate (length (wtFacts wt)) 0
+      ps <- vectorOf (length ts) Zebra.jFactsetId
+      let attribute = Zebra.Attribute (Storable.fromList ts) (Storable.fromList ps) (Storable.fromList tombstones) table
+      entity <- uncurry Zebra.Entity <$> Zebra.jEntityHashId <*> pure (Boxed.singleton attribute)
+      pure . Just $ ZebraWellTyped wt ty entity
+
+-- we are not reading nested arrays in zebra right now
+schemaOfType' :: ValType -> Maybe Schema
+schemaOfType' ty = case ty of
+  ArrayT {} ->
+    Nothing
+  BufT {} ->
+    Nothing
+  MapT {} ->
+    Nothing
+  _ ->
+    schemaOfType ty
+
+schemaOfType :: ValType -> Maybe Schema
+schemaOfType ty = case ty of
+  BoolT ->
+    pure Schema.Bool
+
+  TimeT ->
+    pure Schema.Int
+
+  DoubleT ->
+    pure Schema.Double
+
+  IntT ->
+    pure Schema.Int
+
+  StringT ->
+    pure $ Schema.Array Schema.Byte
+
+  ErrorT ->
+    pure Schema.Int
+
+  UnitT ->
+    Nothing
+
+  FactIdentifierT ->
+    pure Schema.Int
+
+  ArrayT t ->
+    Schema.Array <$> schemaOfType' t
+
+  BufT _ t ->
+    Schema.Array <$> schemaOfType' t
+
+  PairT a b -> do
+    a' <- schemaOfType a
+    b' <- schemaOfType b
+    pure . Schema.Struct . Boxed.fromList $
+      [ Schema.Field (Schema.FieldName "fst") a'
+      , Schema.Field (Schema.FieldName "snd") b' ]
+
+  OptionT t ->
+    let
+      noneOf x =
+        Schema.Variant (Schema.VariantName "none") <$> schemaOfType x
```

Comment:
This should be:
`Schema.Variant (Schema.VariantName "none") (Schema.Struct Boxed.empty)`
or
`Schema.Variant (Schema.VariantName "none") <$> schemaOfType UnitT`